core/arch/arm/plat-hikey/main.c

```diff
@@ -93,3 +108,166 @@ void console_flush(void)
 {
 	pl011_flush(console_base());
 }
+
+static vaddr_t peripherals_base(paddr_t pa)
+{
+	void *va;
+
+	if (cpu_mmu_enabled()) {
+		va = phys_to_virt(pa, MEM_AREA_IO_NSEC);
+		return (vaddr_t)va;
+	}
+	return (vaddr_t)pa;
+}
+
+#ifdef CFG_SPI
+#ifdef CFG_SPI_TEST
+static void spi_test(void)
```

Comment:
Do you mean:
1. You'd rather move the test code into a static TA?
or
2. You'd rather have a 'hook' in the form of a static TA to call `spi_init()`/`spi_test()` in main.c, so that you can test from normal world, rather than test during `driver_init(..)`?
or
3. Something else?

For 1, the concern is that there are things that are specific to hikey only. E.g. register base addresses, clock speed and pin numbers.

For 2, I am testing from xtest currently by hacking core_self_tests.c to `#include "spi.h"` and adding `spi_init()` to `core_self_tests(..)`. Obviously we can create a proper static TA for it (in a later patch?) if deemed necessary, but either way, this is a post-linux test. Calling it from `driver_init(..)` allows for a pre-linux (bootloader stage) test, which is also necessary?